src/sctools/bam.py

```diff
@@ -464,29 +500,83 @@ def __repr__(self) -> str:
+            query_name: str,
+            record: pysam.AlignedSegment = None):
+        self.tag_keys = tag_keys
+        self.tag_values = tag_values
+        self.query_name = query_name
+        self.record = record
+
+    @classmethod
+    def from_aligned_segment(cls, record: pysam.AlignedSegment, tag_keys: List[str] = None) -> 'TagSortableRecord':
+        """Create a ComparableRecord from a pysam.AlignedSegment and list of tag keys"""
+        assert record is not None
+        if tag_keys is None:
+            tag_keys = []
+        tag_values = []
+        for key in tag_keys:
+            tag_value = get_tag_or_default(record, key, '')
+            tag_values.append(tag_value)
+        query_name = record.query_name
+        return cls(tag_keys, tag_values, query_name, record)
+
+    def __lt__(self, other: 'TagSortableRecord'):
+        self.__verify_tag_keys_match(other)
+        for i in range(len(self.tag_keys)):
+            if self.tag_values[i] < other.tag_values[i]:
+                return True
+            elif self.tag_values[i] > other.tag_values[i]:
+                return False
+        return self.query_name < other.query_name
+
+    def __eq__(self, other: 'TagSortableRecord'):
+        self.__verify_tag_keys_match(other)
+        for i in range(len(self.tag_values)):
+            if self.tag_values[i] != other.tag_values[i]:
+                return False
+        return self.query_name == other.query_name
+
+    def __verify_tag_keys_match(self, other):
+        if self.tag_keys != other.tag_keys:
+            format_str = 'Cannot compare records using different tag lists: {0}, {1}'
+            raise ValueError(format_str.format(self.tag_keys, other.tag_keys))
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        format_str = 'ComparableRecord(tags: {0}, tag_values: {1}, query_name: {2}'
+        return format_str.format(self.tag_keys, self.tag_values, self.query_name)
 
-    @property
-    def key_generator(self) -> Callable[[pysam.AlignedSegment], Tuple[str, str, str, str]]:
-        return self._get_sort_key
+    @staticmethod
+    def get_aligned_segments(comparable_records: Iterable['TagSortableRecord']) -> Iterable[pysam.AlignedSegment]:
+        """Get a list of AlignedSegments from a list of ComparableRecords"""
+        aligned_segments = []
+        for cr in comparable_records:
+            if cr.record is None:
+                raise ValueError('AlignedSegment is missing')
+            aligned_segments.append(cr.record)
+        return aligned_segments
 
-    def __repr__(self) -> str:
-        return 'hierarchical__cell_molecule_gene_query_name'
+    @staticmethod
+    def verify_sort(records: Iterable[pysam.AlignedSegment], tag_keys: List[str]) -> None:
+        """Raise AssertionError if the given records are not correctly sorted by the given tags and query name"""
+        old_record = SortFirstTagSortableRecord(tag_keys=tag_keys)
```

Comment:
I think it would be clearer to instantiate a `TagSortableRecord` here directly: `old_record = TagSortableRecord(tag_keys=tag_keys, tag_values=[''] * len(tag_keys), query_name='', record=None)` instead of abstracting it into another class.